zerver/lib/import_realm.py

```diff
@@ -179,16 +179,40 @@ def fix_upload_links(data: TableData, message_table: TableName) -> None:
     organization being imported (which is only determined at import
     time), we need to rewrite the URLs of links to uploaded files
     during the import process.
+
+    Applied to the attachments path_id found in messages of zerver_message and zerver_scheduledmessage tables.
     """
+
+    # No need to loop and check every message if there are no attachments in the first place.
+    if len(path_maps["attachment_path"]) == 0:
+        return
+
     for message in data[message_table]:
         if message["has_attachment"] is True:
-            for key, value in path_maps["attachment_path"].items():
-                if key in message["content"]:
-                    message["content"] = message["content"].replace(key, value)
+            # This more optimized method is used when exporting/importing realm,
+            # this can only be used when we explicitly map each message to its related attachments during export.
+            if "attachments" in message:
+                for attachment_path in message["attachments"]:
+                    message["content"] = message["content"].replace(
+                        attachment_path, path_maps["attachment_path"][attachment_path]
+                    )
                     if message["rendered_content"]:
                         message["rendered_content"] = message["rendered_content"].replace(
-                            key, value
+                            attachment_path, path_maps["attachment_path"][attachment_path]
                         )
+                # Delete attachments key since it's NOT a field in Message or ScheduledMessage.
+                del message["attachments"]
+
+            # Previous approach and is used when a messsage is built from data_import/import_util.build_message()
+            # To do : make the upper approach works for all
+            else:
```

Comment:
We require that exports match the version of the server that they're imported into, so there's no need for a back-compatibility block.